src/sinks/datadog/metrics/sink.rs

```diff
@@ -159,131 +161,72 @@ where
-                {
-                    had_match = true;
-
-                    // Collapse this counter by accumulating its value, and its
-                    // finalizers, and removing it from the original vector of metrics.
-                    accumulated_value += *value;
-
-                    let mut old_metric = metrics.swap_remove(inner_idx);
-                    accumulated_finalizers.merge(old_metric.metadata_mut().take_finalizers());
-                    should_advance = false;
-                } else {
-                    // We hit a counter that _doesn't_ match, but we can't just skip
-                    // it because we also need to evaluate it against all the
-                    // counters that come after it, so we only increment the index
-                    // for this inner loop.
-                    //
-                    // As well, we mark ourselves to stop incrementing the outer
-                    // index if we find more counters to accumulate, because we've
-                    // hit a disjoint counter here. While we may be continuing to
-                    // shrink the count of remaining metrics from accumulating,
-                    // we have to ensure this counter we just visited is visited by
-                    // the outer loop.
-                    is_disjoint = true;
-                }
-            }
-
-            if should_advance {
-                inner_idx += 1;
-
-                if !is_disjoint {
-                    idx += 1;
-                }
-            }
+        if !matches!(right.value(), MetricValue::Counter { .. }) {
+            return false;
+        }
+        if left.series() != right.series() {
+            return false;
         }
 
-        // If we had matches during the accumulator phase, update our original counter.
-        if had_match {
-            let metric = metrics.get_mut(curr_idx).expect("current index must exist");
-            match metric.value_mut() {
-                MetricValue::Counter { value } => {
-                    *value += accumulated_value;
-                    metric
-                        .metadata_mut()
-                        .merge_finalizers(accumulated_finalizers);
-                }
-                _ => unreachable!("current index must represent a counter"),
-            }
+        let left_ts = left.timestamp().map(|dt| dt.timestamp()).unwrap_or(now_ts);
+        let right_ts = right.timestamp().map(|dt| dt.timestamp()).unwrap_or(now_ts);
+        if left_ts != right_ts {
+            return false;
         }
 
-        idx += 1;
-    }
+        // If we reach this point, the counters are for the same series and have the same timestamp.
+        let MetricValue::Counter { value: right_value } = right.value_mut() else {
+            return false;
+        };
+        // NOTE: The docs for `dedup_by` specify that if `left`/`right` are equal, then
+        // `left` is the element that gets removed.
+        *right_value += left_value;
```

Comment:
This doesn't appear to take into account absolute (replace) vs incremental (sum) semantics. Am I missing a step?